doc/computing/randomness.rst

```diff
@@ -0,0 +1,283 @@
+for a discussion.
+
+Using None or `RandomState` instances
+-------------------------------------
+
+Estimators
+..........
+
+As described above, passing instances means that calling `fit()` multiple
+times will not yield the same results, even if the estimator is fitted on the
+same data, with the same hyper-parameters::
+
+    >>> from sklearn.ensemble import RandomForestClassifier
+    >>> from sklearn.datasets import make_classification
+    >>> from sklearn.model_selection import train_test_split
+    >>> import numpy as np
+
+    >>> rng = np.random.RandomState(0)
+    >>> X, y = make_classification(random_state=rng)
+    >>> rf = RandomForestClassifier(max_features=2, max_samples=10,
+    ...                             random_state=rng)
+    >>> X_train, X_test, y_train, y_test = train_test_split(
+    ...     X, y, random_state=rng)
+    >>> rf.fit(X_train, y_train).score(X_test, y_test)
+    0.76
+    >>> rf.fit(X_train, y_train).score(X_test, y_test)
+    0.8
+
+.. note::
+    `score()` is not a random procedure. Only `fit()` has randomness.
+
+We can see from the snippet above that `rf.fit()` has produced different
+models, even if the data was the same. This is because the RNG of the
+estimator is consumed when `fit()` is called, and this consumed (mutaded) RNG
+will be used in the subsequent calls to `fit()`. In addition, the `rng`
+object is shared across any object that uses it, and as a consequence, these
+object become somewhat inter-dependent. For example, two estimator that share
+the same `RandomState` instance will influence each-other, as we will see
+later when we discuss cloning.
+
+If we had passed an int to the `random_state` parameter of the
+:class:`~sklearn.ensemble.RandomForestClassifier`, we would have obtained the
+same models, and thus the same scores each time. When we pass an int, the
+same RNG is used across all calls to `fit()`. What internally happens is that
+even though the RNG is consumed when `fit()` is called, it is always reset to
+its original state at the beginning of `fit()`.
+
+CV splitters
+............
+
+Randomized cv splitters have a similar behavior when a `RandomState`
+instance is passed::
+
+    >>> from sklearn.model_selection import KFold
+    >>> import numpy as np
+    >>> X = np.arange(10)
+    >>> rng = np.random.RandomState(0)
+    >>> cv = KFold(n_splits=2, shuffle=True, random_state=rng)
+    >>> for train, test in cv.split(X):
+    ...     print(train, test)
+    [0 3 5 6 7] [1 2 4 8 9]
+    [1 2 4 8 9] [0 3 5 6 7]
+    >>> for train, test in cv.split(X):
+    ...     print(train, test)
+    [0 4 6 7 8] [1 2 3 5 9]
+    [1 2 3 5 9] [0 4 6 7 8]
+
+We can see that the splits are different from the second time `split()` is
+called. This may lead to wrong results if you compare the performance of
+multiple estimators by calling `split()` many times: the estimators will not
+be evaluated on the same folds, and performances will not be comparable.
+Using an int is usually much safer.
```

Comment:
```suggestion
We can see that the splits are different from the second time `split()` is
called. This may lead to unexpected results if you compare the performance of
multiple estimators by calling `split()` many times: the estimators will not
be evaluated on the same folds, and scores of each fold will not be comparable.
On average, if enough folds are used, one can however expect that that mean
score allows to conclude if one estimator is better than another.
Using an int makes possible score comparisons across folds.
```

Reply:
Would you say that the default (5 folds) is enough so that the average score is still meaningful?

Reply:
it depends on the size of the data. It's a good textbook default.